Initial Kinetis pinctrl Implementation

danieldegrasse opened the pull request:
This PR adds an initial implementation of pinctrl for kinetis series MCUs, with the UART driver enabled. It is based on the work in #39743, but uses a pin group based approach instead of pin nodes. Pin mux configurations are defined at the SOC level as constant values, and these values are referenced in the pin groups in order to select pin mux options.

danieldegrasse committed:
soc: kinetis: Add common pinctrl include file

Add pinctrl include file to define SOC specific pinctrl_soc_t structure,
used to store pin configuration for pinctrl driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pinctrl: Add kinetis pinctrl driver

Add kinetis pinctrl driver. Driver initializes clocks for each port, and
exposes the pinctrl_configure_pins function required for pinctrl
support.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: bindings: Add pinctrl node group binding for kinetis

Add pinctrl node group bindings for kinetis MCUs

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: serial: Add pinctrl support to mcux uart driver

Add pinctrl support to mcux uart driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_k64f: Enable pinctrl

Add board level pin group definitions, and remove uart pinmux
driver calls.

pinctrl is enabled for uart driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

gmarull commented on soc/arm/nxp_kinetis/CMakeLists.txt:
```diff
@@ -4,6 +4,9 @@ zephyr_sources_ifdef(CONFIG_KINETIS_FLASH_CONFIG	flash_configuration.c)
 
 add_subdirectory(${SOC_SERIES})
 
+# This is for access to pinmux macros
```
nit: `pinctrl macros`

gmarull commented on drivers/pinctrl/pinctrl_kinetis.c:
```diff
@@ -0,0 +1,69 @@
+			DT_INST_CLOCKS_CELL(n, bits))
+#endif
+
+#define PINMUX_MCUX_INIT(n)						\
```
nit: please use `PINCTRL` to avoid confusion with old pinmux API.

gmarull commented on dts/bindings/pinctrl/nxp,kinetis-pinctrl.yaml:
```diff
@@ -0,0 +1,78 @@
+    description: |
+      Kinetis pin controller pin configuration node
+    properties:
+      pins:
```
nit: should likely be named `pinmux`, since this field describes the muxing right?

gmarull commented on dts/arm/nxp/nxp_k6x.dtsi:
```diff
@@ -71,6 +71,14 @@
 	};
 
+
+	/* Dummy pinctrl node, filled with pin mux options at board leve; */
```
typo: `level`.

gmarull commented on drivers/serial/uart_mcux.c:
```diff
@@ -358,11 +369,20 @@ static const struct uart_driver_api uart_mcux_driver_api = {
 
+#ifdef CONFIG_PINCTRL
+#define PINCTRL_INIT(n) .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),
+#define PINCTRL_DECLARE(n) PINCTRL_DT_INST_DEFINE(n);
```
nit: `_DEFINE`

gmarull commented on boards/arm/frdm_k64f/frdm_k64f-pinctrl.dtsi:
```diff
@@ -0,0 +1,130 @@
+&pinctrl {
+	i2c0_default: i2c0_default {
+		group0 {
+			pins = <&I2C0_SCL_PTE24
+				&I2C0_SDA_PTE25>;
```
note: I think `lower_case` nodes would align better with DT style. `UPPER_CASE` is usually for macros/consts. @mbolivar-nordic can you comment on this one please?

danieldegrasse replied:
I agree here that lower case nodes would be best, the reason for these node names is that NXP's HAL already has defined pinmux options for Kinetis parts using this style, so those files would have to be regenerated with lowercase node naming. (See the K64 pinmux defintion [here](https://github.com/zephyrproject-rtos/hal_nxp/blob/master/dts/nxp/kinetis/MK64FN1M0VLL12-pinctrl.dtsi))

mbolivar-nordic replied:
Right, galak did it this way if I recall correctly, so the ship has sailed.

gmarull replied:
@danieldegrasse it looks like you can just use definitions instead of nodes for this case (seems like 32-bit should be enough to store stuff like `nxp,kinetis-port-pins = < 13 2 >;`.

danieldegrasse replied:
You definitely could- the two values combined will fit into 32 bits easily. The existing pinctrl definitions in NXP's HAL would have to be regenerated, but there's no reason that can't be done. @mbolivar-nordic, do you think regenerating the pinctrl definitions avoid using nodes here would be best? The final implementation would likely end up looking similar to the one introduced in #39974. The main benefits would be slightly smaller code size (since the `pinctrl_soc_t` typedef would be a `uint32_t`), and better alignment with the pinctrl implementation guide.

gmarull replied:
I think that using definitions would be better here, nodes are not really required in this case. AFAIK, the generated nodes have never been used in reality, so it's a good time to make this kind of decision.

gmarull reviewed (commented):
Overall LGTM, I have some minor comments

gmarull commented on drivers/pinctrl/pinctrl_kinetis.c:
```diff
@@ -0,0 +1,80 @@
+	return 0;
+}
+
+/* Kinetis pinmux driver binds to the same DTS nodes,
+ * and handles clock init. Only bind to these nodes if pinmux driver
+ * is disabled.
+ */
+#ifndef CONFIG_PINMUX
```
note: are you planning to support `CONFIG_PINMUX=y` + `CONFIG_PINCTRL=y`?

gmarull commented on dts/bindings/pinctrl/nxp,kinetis-pinctrl.yaml:
```diff
@@ -0,0 +1,78 @@
+
+  uart0_default: uart0_default {
+    group0 {
+      pins = <UART0_RX_PTB16
+        UART0_TX_PTB17>;
```
nit: `pinmux = <UART0_RX_PTB16>, <UART0_TX_PTB17>;`

gmarull commented on boards/arm/frdm_k64f/frdm_k64f-pinctrl.dtsi:
```diff
@@ -0,0 +1,130 @@
+&pinctrl {
+	i2c0_default: i2c0_default {
+		group0 {
+			pinmux = <I2C0_SCL_PTE24
+				I2C0_SDA_PTE25>;
```
nit: `pinmux = <I2C0_SCL_PTE24>, <I2C0_SDA_PTE25>;` (should be equivalent, but it's the format found in the pinctrl guide, also used by STM32 in Linux Kernel).

gmarull commented on boards/arm/frdm_k64f/frdm_k64f_defconfig:
```diff
@@ -7,6 +7,7 @@ CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
 CONFIG_CORTEX_M_SYSTICK=y
 CONFIG_GPIO=y
+CONFIG_PINCTRL=y
 CONFIG_PINMUX=y
```
same question: are both needed?

danieldegrasse replied:
see response above

gmarull reviewed (dismissed):
LGTM, I have a few minor questions/suggestions.

danieldegrasse committed:
boards: frdm_kw41z: Add pinctrl nodes for frdm_kw41z

Add pinctrl nodes for frdm_kw41z

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: hexiwear_k64: add pinctrl groups for hexiwear_k64

Add pinctrl groups for hexiwear_k64, and enable pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: hexiwear_kw40z: Enable pinctrl for kw40z

Enable pinctrl for hexiwear kw40z target

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: ip_k66f: Add pinctrl groups for ip_k66f

Add pinctrl groups for ip_k66f

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_k22f: Add pinctrl nodes for frdm_k22f

Add pinctrl nodes for frdm_k22f

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_k82f: Add pinctrl groups for frdm_k82f

Add pinctrl groups for frdm_k82f

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: frdm_kl25z: Add pinctrl groups for frdm_kl25z

Add pinctrl groups for frdm_kl25z

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: nxp: Add pinctrl dummy node for kinetis series MCUs

Kinetis pinctrl groups need dummy pinctrl node to populate with pinctrl
options at the board level. Add one in each

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: nxp: Remove SOC specific wrapper definitions

SOC specific wrapper files were used to include pinctrl DTSI files.
Remove these in favor of including pinctrl header files directly from
board pinctrl definitions.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
west.yml: Update nxp hal to use new pinctrl header files

NXP hal pinctrl header files are required for pin groups. Update HAL
revision to pull in pinctrl header files.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: rddrone_fmuk66: Add pinctrl groups for rddrone_fmuk66

Add pinctrl groups for rddrone_fmuk66

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: twr_ke18f: Add pinctrl groups for twr_ke18f

Add pinctrl groups for twr_ke18f

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: twr_kv58f220m: Add pinctrl groups for twr_kv58f220m

Add pinctrl groups for twr_kv58f220m

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: usb_kw24d512: Add pinctrl groups for usb_kw24d512

Add pinctrl groups for usb_kw24d512

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic commented:
@gmarull please re-approve

mbolivar-nordic commented:
@dleach02 @mmahadevan108 please review
